### PR TITLE
Cleanup oadm router somewhat

### PR DESCRIPTION
**Pull request by smarterclayton**
Still more work to do in the router command, it's gotten... opaque.

@liggitt

### CODE REVIEW

**Comment by smarterclayton** [comment]
@detiber starting with 1.2/3.2 we will want to pass --subdomain=ROUTER_SUBDOMAIN to oadm router in a single le.  The default config should remain for back compat.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Ok, this now includes even more function - it:

* remove hostport support from oadm router - it's useless and someone can set that on their own
* store default certificate as a secret and plumb the path all the way down to the haproxy instance
* change the default from using ENV for secrets to using service account, and generate resources for each thing that needs to be set
  * default certificate is passed as bytes if this flag is set, otherwise is set as a path environment variable
* cleanup the SCC warning on the router
* the --ports mapping is now repurposed to the mapping on the router service (to make routers easier to remap)
* support --subdomain as a template

I've done local testing and everything seems to be ok, but needs a bit more look and love.

@detiber FYI this is changing router default to create a service account - all other flags are the same, but you can now omit --credentials, and we probably want to start setting --subdomain

@pweil- @ramr for review

**Comment by smarterclayton** [comment]
[extended:core(router)]

File: pkg/cmd/admin/router/router.go [hunk 1]
```diff
@@ -64,20 +66,31 @@ you have failover protection.`
 	secretsVolumeName = "secret-volume"
 	secretsPath       = "/etc/secret-volume"
 
+	defaultCertificateDir = "/etc/pki/tls/private"
```
Comment (liggitt) [inline pkg/cmd/admin/router/router.go]: is this standard? if so, `[citation needed]` as a comment

Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: It's the RH standard, vs ubuntu which is using the older /etc/ssl

**Comment by smarterclayton** [comment]
@ramr review please

**Comment by smarterclayton** [comment]
moar eyes

**Comment by pweil-** [comment]
@kargakis - would you mind giving this some eyeballs since I'm tied up still?

File: pkg/cmd/admin/router/router.go [hunk 2]
```diff
@@ -288,15 +298,45 @@ func generateSecretsConfig(cfg *RouterConfig, kClient *kclient.Client,
...
+	}
+
+	if len(defaultCert) > 0 {
+		// TODO: extract the private key from the CRT and set it as its own key
+		// TODO: use upstream constants for this
+		secret := &kapi.Secret{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: fmt.Sprintf("%s-certs", cfg.Name),
+			},
+			Type: "kubernetes.io/tls",
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: https://github.com/kubernetes/kubernetes/blob/0866c377e9a84a73e0ff78efdd1def7f7e8b437a/pkg/api/types.go#L2311

Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: Not in our branch yet

On Wed, Feb 17, 2016 at 11:51 AM, Michail Kargakis <notifications@github.com

File: pkg/cmd/admin/router/router.go [hunk 3]
```diff
@@ -469,26 +507,29 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
+	if len(cfg.ServiceAccount) == 0 {
+		return fmt.Errorf("you must specify a service account for the router with --service-account")
+	}
 
-		// create new router
-		if len(cfg.Credentials) == 0 {
-			return fmt.Errorf("router could not be created; you must specify a .kubeconfig file path containing credentials for connecting the router to the master with --credentials")
-		}
+	if err := validateServiceAccount(kClient, namespace, cfg.ServiceAccount, cfg.HostNetwork); err != nil {
+		return fmt.Errorf("router could not be created; %v", err)
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: s/;/:/

File: pkg/cmd/admin/router/router.go [hunk 4]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
+	if output {
+		list.Items, err = ocmdutil.ConvertItemsForDisplayFromDefaultCommand(cmd, list.Items)
+		if err != nil {
+			return err
 		}
-		if errs := bulk.Create(list, namespace); len(errs) != 0 {
-			return errExit
+
+		if err := f.PrintObject(cmd, list, out); err != nil {
+			return fmt.Errorf("Unable to print object: %v", err)
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: lowercase

File: pkg/cmd/admin/router/router.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: This was intentional, the error may not be related to the failure

On Wed, Feb 17, 2016 at 12:14 PM, Michail Kargakis <notifications@github.com

File: pkg/cmd/admin/router/router.go [hunk 5]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
+	mapper, typer := f.Factory.Object()
+	bulk := configcmd.Bulk{
+		Mapper:            mapper,
+		Typer:             typer,
+		RESTClientFactory: f.Factory.ClientForMapping,
+
+		After: configcmd.NewPrintNameOrErrorAfter(mapper, cmdutil.GetFlagString(cmd, "output") == "name", "created", out, cmd.Out()),
+	}
+	if errs := bulk.Create(list, namespace); len(errs) != 0 {
+		return errExit
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: Is errExit more informative than errs aggregated?

File: pkg/cmd/admin/router/router.go [hunk 6]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
-		if len(secrets) != 0 {
-			serviceAccount, err := kClient.ServiceAccounts(namespace).Get(cfg.ServiceAccount)
-			if err != nil {
-				return fmt.Errorf("error looking up service account %s: %v",
-					cfg.ServiceAccount, err)
-			}
-
-			for _, secret := range secrets {
-				objects = append(objects, secret)
+	updatePercent := int(-25)
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: can you move this just before creating the DC?

File: pkg/cmd/admin/router/router.go [hunk 7]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
 
-				serviceAccount.Secrets = append(serviceAccount.Secrets,
-					kapi.ObjectReference{Name: secret.Name})
-			}
+	secrets, volumes, mounts, err := generateSecretsConfig(cfg, kClient,
+		namespace, defaultCert)
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: I am pretty sure this would be fine in one line

File: pkg/cmd/admin/router/router.go [hunk 8]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
+				Subjects: []kapi.ObjectReference{
+					{
+						Kind:      "ServiceAccount",
+						Name:      cfg.ServiceAccount,
+						Namespace: namespace,
+					},
+				},
+				RoleRef: kapi.ObjectReference{
+					Kind: "ClusterRole",
+					Name: "system:router",
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: https://github.com/openshift/origin/blob/d1a5a27d911e7836229992d128afa6d16dde30f4/pkg/cmd/server/bootstrappolicy/constants.go#L67

File: pkg/cmd/admin/router/router.go [hunk 9]
```diff
@@ -505,154 +546,183 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, cfg *
...
+	})
+
+	objects = app.AddServices(objects, false)
+	// set the service port to the provided hostport value
+	for i := range objects {
+		switch t := objects[i].(type) {
+		case *kapi.Service:
+			for j, servicePort := range t.Spec.Ports {
+				for _, targetPort := range ports {
+					if targetPort.ContainerPort == servicePort.Port && targetPort.HostPort != 0 {
```
Comment (kargakis) [inline pkg/cmd/admin/router/router.go]: Shouldn't this compare the container port against the service target port?

File: pkg/cmd/admin/router/router.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: we print in order of user input here, vs returning user successes then
failures.  I like interspersed execution better.

On Wed, Feb 17, 2016 at 12:29 PM, Michail Kargakis <notifications@github.com

File: pkg/cmd/admin/router/router.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: I don't like constant references that aren't part of the API being taken
across package groups.  This has to be very deliberately changed.

On Wed, Feb 17, 2016 at 12:59 PM, Michail Kargakis <notifications@github.com

File: pkg/cmd/admin/router/router.go (hunk 9 above)
Comment (smarterclayton) [inline pkg/cmd/admin/router/router.go]: HostPort is "external", ContainerPort is "internal", so the service port is
going to be the "external" port.  That's roughly equivalent to how host
ports work, which is the closest intent when you aren't using networks.

File: pkg/cmd/admin/router/router.go (hunk 8 above)
Comment (liggitt) [inline pkg/cmd/admin/router/router.go]: > I don't like constant references that aren't part of the API being taken across package groups.  This has to be very deliberately changed.

agree

**Comment by smarterclayton** [comment]
@liggitt review last commit

**Comment by liggitt** [comment]
one question on newline in encoding PEM blocks

**Comment by smarterclayton** [comment]
Does add trailing newline.

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Well this is a nice flake:

```
ERROR: gave up waiting for https://127.0.0.1:38443/healthz

!!! Error in /data/src/github.com/openshift/origin/hack/../hack/util.sh:370
	'return 1' exited with status 1
Call stack:
	1: /data/src/github.com/openshift/origin/hack/../hack/util.sh:370 start_os_master(...)
	2: /data/src/github.com/openshift/origin/hack/update-generated-swagger-spec.sh:50 main(...)
Exiting with status 1
[INFO] Dumping container logs to /tmp/openshift/generate-swagger-spec//logs
[INFO] Dumping all resources to /tmp/openshift/generate-swagger-spec//logs/export_all.json
Authentication required for https://127.0.0.1:38443 (openshift)
Username: system:admin
Password: error: username system:admin is invalid for basic auth
error: failed to negotiate an api version; server supports: map[], client supports: map[v1:{} v1beta3:{} extensions/v1beta1:{} metrics/v1alpha1:{} componentconfig/v1alpha1:{} authorization.k8s.io/v1beta1:{}]
```

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
```
Error from server: buildconfigs "test" not found
W0219 00:44:52.316886    5128 cmd.go:212] -p POD is DEPRECATED and will be removed in a future version. Use exec POD instead.
!!! Error in hack/../hack/../test/end-to-end/core.sh:299
	'grep -q "unable to validate against any security context constraint"' exited with status 0 1
Call stack:
	1: hack/../hack/../test/end-to-end/core.sh:299 main(...)
Exiting with status 1
Error from server: User "e2e-default-admin" cannot list all buildconfigs in the cluster
```

[test]

**Pull request merged**